chia/server/rate_limits.py

```diff
@@ -129,31 +146,49 @@ def process_msg_and_check(self, message: Message) -> bool:
-        self.message_cumulative_sizes[message_type] += len(message.data)
-        proportion_of_limit = self.percentage_of_limit / 100
-
-        limits = DEFAULT_SETTINGS
-        if message_type in rate_limits_tx:
-            limits = rate_limits_tx[message_type]
-        elif message_type in rate_limits_other:
-            limits = rate_limits_other[message_type]
-            self.non_tx_message_counts += 1
-            self.non_tx_cumulative_size += len(message.data)
-            if self.non_tx_message_counts > NON_TX_FREQ * proportion_of_limit:
+        new_message_counts: int = self.message_counts[message_type] + 1
+        new_cumulative_size: int = self.message_cumulative_sizes[message_type] + len(message.data)
+        new_non_tx_count: int = self.non_tx_message_counts
+        new_non_tx_size: int = self.non_tx_cumulative_size
+        proportion_of_limit: float = self.percentage_of_limit / 100
+
+        ret: bool = False
+        try:
+
+            limits = DEFAULT_SETTINGS
+            if message_type in rate_limits_tx:
+                limits = rate_limits_tx[message_type]
+            elif message_type in rate_limits_other:
+                limits = rate_limits_other[message_type]
+                new_non_tx_count = self.non_tx_message_counts + 1
+                new_non_tx_size = self.non_tx_cumulative_size + len(message.data)
+                if new_non_tx_count > NON_TX_FREQ * proportion_of_limit:
+                    return False
+                if new_non_tx_size > NON_TX_MAX_TOTAL_SIZE * proportion_of_limit:
+                    return False
+            else:
+                log.warning(f"Message type {message_type} not found in rate limits")
+
+            if limits.max_total_size is None:
+                limits = dataclasses.replace(limits, max_total_size=limits.frequency * limits.max_size)
+            assert limits.max_total_size is not None
+
+            if new_message_counts > limits.frequency * proportion_of_limit:
+                return False
+            if len(message.data) > limits.max_size:
                 return False
-            if self.non_tx_cumulative_size > NON_TX_MAX_TOTAL_SIZE * proportion_of_limit:
+            if new_cumulative_size > limits.max_total_size * proportion_of_limit:
                 return False
-        else:
-            log.warning(f"Message type {message_type} not found in rate limits")
-
-        if limits.max_total_size is None:
-            limits = dataclasses.replace(limits, max_total_size=limits.frequency * limits.max_size)
-
-        if self.message_counts[message_type] > limits.frequency * proportion_of_limit:
-            return False
-        if len(message.data) > limits.max_size:
-            return False
-        if self.message_cumulative_sizes[message_type] > limits.max_total_size * proportion_of_limit:
-            return False
-        return True
+
+            ret = True
+            return True
+        finally:
+            if self.incoming or ret:
```

Comment:
only commit the new counters if we're returning `True` or if this was an incoming rate limiter, and we already received the message. In the latter case we should always increment the counters.